### DISCUSSION TITLE
Normal and motion mode keybindings and input methods

### DISCUSSION THREAD

**Discussion by algor512** [meow-edit/meow#274] [General]
Is it possible to make the keybindings work in non-english input methods, like russian-computer?

**Comment by eshrh** [comment]
Sure, it'll just take some effort from you to make your own keymap. As far as I can tell, any symbol can be bound with the  `(meow-normal-define-key)` helper. You might be able to transpose the suggested qwerty keybindings onto the russian-computer input method, but all you need to do is choose which key will be bound to which command. Additionally, any number of keys can be bound to the same command.

**Reply by algor512** [reply]
Thank you for the reply!

It is a solution, indeed, but it may cause a problem when e.g. I wish to bind something different on `?` and on `,`, as `?` in the English input method is `,` in the Russian one.

**Reply by eshrh** [reply]
hmm, that's interesting. meow keybindings are not complicated, and they're really just emacs keymaps internally. I guess what you could do is, write a function (like meow-setup) that sets up your english layout keybindings, and then write a function that sets up your russian layout keybindings, and then you can do something like

```elisp
(add-hook 'input-method-activate-hook
          (lambda ()
            (if (string= current-input-method "russian-computer")
                (meow-setup-russian-computer)
              (meow-setup-english))))
```

to be honest, i'm not sure if it's possible to get what you're describing built in, because the meow states don't have much abstraction above simple emacs minor modes and keymaps.

**Reply by algor512** [reply]
Yes, it's strange. I was playing a bit with `define-key`, and noticed that bindings with modifier keys (`C-<letter>`) work well in the Russian input method, but one-letter bindings break in it.

I also noticed that Evil-mode bindings work well in both input methods. Perhaps Evil uses something like what you propose. Maybe I'll try to understand how it works there and use something similar in my meow config.

**Reply by algor512** [reply]
I tried modalka-mode (https://github.com/mrkkrp/modalka), and there one-letter bindings work well. I guess that the reason why is here:
```
(advice-add 'quail-input-method :around #'modalka--input-function-advice)
```

But I am not sure as I hardly understand what `quail-input-method` does.

**Comment by eshrh** [comment]
(just want to move out of the long reply thread)

can you give a step by step process to reproduce this and give the expected behavior? I'm not totally clear on what you're talking about, but if it works in evil and modalka then maybe it's something we can implement.

**Reply by algor512** [reply]
Ok, I've done this - https://github.com/meow-edit/meow/issues/277
